sos/plugins/nvme.py

```diff
@@ -0,0 +1,119 @@
+
+            queue_files = subprocess.check_output("ls -1 /sys/block/nvme%dn1/queue/*" % i, shell=True).rstrip()
+            # in python 3.3 or above the prefix b' is not ignored, so bytes.decode() is used to get rid of it
+            queue_files = bytes.decode(queue_files)
+            queue_files = queue_files.splitlines()
+            for q in queue_files:
+                self.add_copy_spec(q)
+
+            device_files = subprocess.check_output("find /sys/block/nvme%dn1/device/ -maxdepth 1 -type f" % i, shell=True).rstrip()
+            device_files = bytes.decode(device_files)
+            device_files = device_files.splitlines()
+            for d in device_files:
+                self.add_copy_spec(d)
+
+            general_files = subprocess.check_output("find /sys/block/nvme%dn1/ -maxdepth 1 -type f" % i, shell=True).rstrip()
+            general_files = bytes.decode(general_files)
+            general_files = general_files.splitlines()
+            for g in general_files:
+                self.add_copy_spec(g)
+
+            integrity_files = subprocess.check_output("ls -1 /sys/block/nvme%dn1/integrity/*" % i, shell=True).rstrip()
+            integrity_files = bytes.decode(integrity_files)
+            integrity_files = integrity_files.splitlines()
+            for ing in integrity_files:
+                self.add_copy_spec(ing)
+
+            mq_files = subprocess.check_output("find /sys/block/nvme%dn1/mq/ -type f" % i, shell=True).rstrip()
+            mq_files = bytes.decode(mq_files)
+            mq_files = mq_files.splitlines()
+            for m in mq_files:
+                self.add_copy_spec(m)	
+
+            i += 1
+
+    def setup(self):
+        self.copy_spec()
+
+        for dev in self.devices:
+            # get block size
+            self.add_cmd_output("sh -c \"lsblk | grep %s | awk '{ print $4 }'\"" % dev, suggest_filename="block-size.%s" % dev)
+
+            # check if the firmware mode is OPAL
+            opal = subprocess.check_output("cat /proc/cpuinfo | grep firmware | grep OPAL | wc -l", shell=True)
+            opal = int(opal)
+            if opal == 1:
+                grep_op = "mass-storage"
+            else:
+                grep_op = "pci"
+
+            # get info about slot location and pci location
+            slot = subprocess.check_output("lscfg -vl %s | grep nvme | grep %s | awk '{ print $4 }'"
+                % (dev[0:-2], grep_op), shell=True).strip()
+            pci = subprocess.check_output("lscfg -vl %s | grep nvme | grep %s | awk '{ print $1 }'" 
+                % (dev[0:-2], grep_op), shell=True).strip()
+
+            # add slot and pci location in sos_strings/nvme/
+            self.add_string_as_file(slot, "slot_location.%s" % dev)
+            self.add_string_as_file(pci, "pci_location.%s" % dev)
+
+            # get PCI Vendor ID
+            self.add_cmd_output("sh -c \"lspci -vs %s | grep 'Non-Volatile memory controller' | cut -c 46-\"" 
```

Comment:
No pipelines. Also any magic numbers or strings like `46-` deserve an explanation in a comment.